File: ios/runanywhere-sdks-latest/examples/ios/RunAnywhereAI/RunAnywhereAI/Features/Benchmarks/Services/BenchmarkRunner.swift
//
//  BenchmarkRunner.swift
//  RunAnywhereAI
//
//  Orchestrates benchmark execution across all providers.
//

import Foundation
import RunAnywhere

// MARK: - Provider Protocol

protocol BenchmarkScenarioProvider: Sendable {
    var category: BenchmarkCategory { get }
    func scenarios() -> [BenchmarkScenario]
    func execute(
        scenario: BenchmarkScenario,
        model: ModelInfo
    ) async throws -> BenchmarkMetrics
}

// MARK: - Runner Errors

enum BenchmarkRunnerError: LocalizedError {
    case noModelsAvailable(skippedCategories: [BenchmarkCategory])
    case fetchModelsFailed(underlying: Error)

    var errorDescription: String? {
        switch self {
        case .noModelsAvailable(let skipped):
            let names = skipped.map(\.displayName).joined(separator: ", ")
            return "No downloaded models found for: \(names). Download models first from the Models tab."
        case .fetchModelsFailed(let error):
            return "Failed to fetch available models: \(error.localizedDescription)"
        }
    }
}

// MARK: - Pre-flight Result

struct BenchmarkPreflightResult: Sendable {
    let availableCategories: [BenchmarkCategory: [ModelInfo]]
    let skippedCategories: [BenchmarkCategory]
    let totalWorkItems: Int
}

// MARK: - Runner

final class BenchmarkRunner {

    private let providers: [BenchmarkCategory: BenchmarkScenarioProvider]

    init() {
        var map: [BenchmarkCategory: BenchmarkScenarioProvider] = [:]
        let all: [BenchmarkScenarioProvider] = [
            LLMBenchmarkProvider(),
            STTBenchmarkProvider(),
            TTSBenchmarkProvider(),
            VLMBenchmarkProvider(),
            DiffusionBenchmarkProvider(),
        ]
        for provider in all {
            map[provider.category] = provider
        }
        self.providers = map
    }

    // MARK: - Preflight Check

    /// Checks which categories have downloaded models before running. This lets the UI
    /// inform the user which categories will be skipped.
    func preflight(categories: Set<BenchmarkCategory>) async throws -> BenchmarkPreflightResult {
        let allModels: [ModelInfo]
        do {
            allModels = try await RunAnywhere.availableModels()
        } catch {
            throw BenchmarkRunnerError.fetchModelsFailed(underlying: error)
        }

        var available: [BenchmarkCategory: [ModelInfo]] = [:]
        var skipped: [BenchmarkCategory] = []

        for category in BenchmarkCategory.allCases where categories.contains(category) {
            guard providers[category] != nil else {
                skipped.append(category)
                continue
            }
            let models = allModels.filter {
                $0.category == category.modelCategory && $0.isDownloaded && !$0.isBuiltIn
            }
            if models.isEmpty {
                skipped.append(category)
            } else {
                available[category] = models
            }
        }

        var totalItems = 0
        for (category, models) in available {
            let scenarioCount = providers[category]?.scenarios().count ?? 0
            totalItems += models.count * scenarioCount
        }

        return BenchmarkPreflightResult(
            availableCategories: available,
            skippedCategories: skipped,
            totalWorkItems: totalItems
        )
    }

    // MARK: - Run

    func runBenchmarks(
        categories: Set<BenchmarkCategory>,
        onProgress: @escaping @Sendable (BenchmarkProgressUpdate) -> Void
    ) async throws -> BenchmarkRunOutput {
        let preflight = try await preflight(categories: categories)

        // If nothing to run, throw a descriptive error
        if preflight.availableCategories.isEmpty {
            throw BenchmarkRunnerError.noModelsAvailable(
                skippedCategories: preflight.skippedCategories
            )
        }

        // Build work list: (category, model, scenario)
        var workItems: [(BenchmarkCategory, ModelInfo, BenchmarkScenario)] = []

        for category in BenchmarkCategory.allCases where categories.contains(category) {
            guard let provider = providers[category],
                  let models = preflight.availableCategories[category] else { continue }
            let scenarioList = provider.scenarios()
            for model in models {
                for scenario in scenarioList {
                    workItems.append((category, model, scenario))
                }
            }
        }

        let total = workItems.count
        var results: [BenchmarkResult] = []

        for (index, (category, model, scenario)) in workItems.enumerated() {
            try Task.checkCancellation()

            onProgress(BenchmarkProgressUpdate(
                completedCount: index,
                totalCount: total,
                currentScenario: scenario.name,
                currentModel: model.name
            ))

            let metrics: BenchmarkMetrics
            do {
                guard let provider = providers[category] else { continue }
                metrics = try await provider.execute(
                    scenario: scenario,
                    model: model
                )
            } catch is CancellationError {
                throw CancellationError()
            } catch {
                var errorMetrics = BenchmarkMetrics()
                errorMetrics.errorMessage = "\(category.displayName) [\(model.name)]: \(error.localizedDescription)"
                metrics = errorMetrics
            }

            results.append(BenchmarkResult(
                category: category,
                scenario: scenario,
                modelInfo: ComponentModelInfo(from: model),
                metrics: metrics
            ))
        }

        // Final progress
        onProgress(BenchmarkProgressUpdate(
            completedCount: total,
            totalCount: total,
            currentScenario: "Done",
            currentModel: ""
        ))

        return BenchmarkRunOutput(
            results: results,
            skippedCategories: preflight.skippedCategories
        )
    }
}

File: ios/runanywhere-sdks-latest/examples/android/RunAnywhereAI/app/src/main/java/com/runanywhere/runanywhereai/presentation/benchmarks/services/BenchmarkRunner.kt
package com.runanywhere.runanywhereai.presentation.benchmarks.services

import com.runanywhere.runanywhereai.presentation.benchmarks.models.BenchmarkCategory
import com.runanywhere.runanywhereai.presentation.benchmarks.models.BenchmarkDeviceInfo
import com.runanywhere.runanywhereai.presentation.benchmarks.models.BenchmarkMetrics
import com.runanywhere.runanywhereai.presentation.benchmarks.models.BenchmarkProgressUpdate
import com.runanywhere.runanywhereai.presentation.benchmarks.models.BenchmarkResult
import com.runanywhere.runanywhereai.presentation.benchmarks.models.BenchmarkScenario
import com.runanywhere.runanywhereai.presentation.benchmarks.models.ComponentModelInfo
import com.runanywhere.runanywhereai.presentation.benchmarks.utilities.SyntheticInputGenerator
import com.runanywhere.sdk.public.RunAnywhere
import com.runanywhere.sdk.public.extensions.Models.ModelInfo
import com.runanywhere.sdk.public.extensions.availableModels
import kotlinx.coroutines.ensureActive
import kotlin.coroutines.coroutineContext

// -- Provider Interface --

interface BenchmarkScenarioProvider {
    val category: BenchmarkCategory
    fun scenarios(): List<BenchmarkScenario>
    suspend fun execute(
        scenario: BenchmarkScenario,
        model: ModelInfo,
        deviceInfo: BenchmarkDeviceInfo,
    ): BenchmarkMetrics
}

// -- Runner Errors --

sealed class BenchmarkRunnerError(message: String) : Exception(message) {
    class NoModelsAvailable(skippedCategories: List<BenchmarkCategory>) : BenchmarkRunnerError(
        "No downloaded models found for: ${skippedCategories.joinToString { it.displayName }}. Download models first from the Models tab.",
    )

    class FetchModelsFailed(cause: Throwable) : BenchmarkRunnerError(
        "Failed to fetch available models: ${cause.localizedMessage ?: cause.message ?: "Unknown error"}",
    )
}

// -- Preflight Result --

data class BenchmarkPreflightResult(
    val availableCategories: Map<BenchmarkCategory, List<ModelInfo>>,
    val skippedCategories: List<BenchmarkCategory>,
    val totalWorkItems: Int,
)

// -- Runner --

class BenchmarkRunner {

    private val providers: Map<BenchmarkCategory, BenchmarkScenarioProvider>

    init {
        val all = listOf(
            LLMBenchmarkProvider(),
            STTBenchmarkProvider(),
            TTSBenchmarkProvider(),
            VLMBenchmarkProvider(),
        )
        providers = all.associateBy { it.category }
    }

    // -- Preflight Check --

    suspend fun preflight(categories: Set<BenchmarkCategory>): BenchmarkPreflightResult {
        val allModels: List<ModelInfo> = try {
            RunAnywhere.availableModels()
        } catch (e: Exception) {
            throw BenchmarkRunnerError.FetchModelsFailed(e)
        }

        val available = mutableMapOf<BenchmarkCategory, List<ModelInfo>>()
        val skipped = mutableListOf<BenchmarkCategory>()

        for (category in BenchmarkCategory.entries) {
            if (category !in categories) continue
            if (providers[category] == null) {
                skipped.add(category)
                continue
            }
            val models = allModels.filter {
                it.category == category.modelCategory && it.isDownloaded && !it.isBuiltIn
            }
            if (models.isEmpty()) {
                skipped.add(category)
            } else {
                available[category] = models
            }
        }

        var totalItems = 0
        for ((category, models) in available) {
            val scenarioCount = providers[category]?.scenarios()?.size ?: 0
            totalItems += models.size * scenarioCount
        }

        return BenchmarkPreflightResult(
            availableCategories = available,
            skippedCategories = skipped,
            totalWorkItems = totalItems,
        )
    }

    // -- Run Result (includes skipped categories) --

    data class BenchmarkRunResult(
        val results: List<BenchmarkResult>,
        val skippedCategories: List<BenchmarkCategory>,
    )

    // -- Run --

    suspend fun runBenchmarks(
        categories: Set<BenchmarkCategory>,
        onProgress: (BenchmarkProgressUpdate) -> Unit,
    ): BenchmarkRunResult {
        val preflightResult = preflight(categories)

        if (preflightResult.availableCategories.isEmpty()) {
            throw BenchmarkRunnerError.NoModelsAvailable(preflightResult.skippedCategories)
        }

        data class WorkItem(
            val category: BenchmarkCategory,
            val model: ModelInfo,
            val scenario: BenchmarkScenario,
        )

        val workItems = mutableListOf<WorkItem>()
        for (category in BenchmarkCategory.entries) {
            if (category !in categories) continue
            val provider = providers[category] ?: continue
            val models = preflightResult.availableCategories[category] ?: continue
            val scenarioList = provider.scenarios()
            for (model in models) {
                for (scenario in scenarioList) {
                    workItems.add(WorkItem(category, model, scenario))
                }
            }
        }

        val total = workItems.size
        val results = mutableListOf<BenchmarkResult>()

        for ((index, item) in workItems.withIndex()) {
            coroutineContext.ensureActive()

            onProgress(
                BenchmarkProgressUpdate(
                    completedCount = index,
                    totalCount = total,
                    currentScenario = item.scenario.name,
                    currentModel = item.model.name,
                ),
            )

            val metrics: BenchmarkMetrics = try {
                val provider = providers[item.category] ?: continue
                provider.execute(
                    scenario = item.scenario,
                    model = item.model,
                    deviceInfo = BenchmarkDeviceInfo(
                        modelName = "",
                        chipName = "",
                        totalMemoryBytes = 0,
                        availableMemoryBytes = SyntheticInputGenerator.availableMemoryBytes(),
                        osVersion = "",
                    ),
                )
            } catch (e: kotlinx.coroutines.CancellationException) {
                throw e
            } catch (e: Exception) {
                BenchmarkMetrics(
                    errorMessage = "${item.category.displayName} [${item.model.name}]: ${e.localizedMessage ?: e.message ?: "Unknown error"}",
                )
            }

            results.add(
                BenchmarkResult(
                    category = item.category,
                    scenario = item.scenario,
                    modelInfo = ComponentModelInfo.from(item.model),
                    metrics = metrics,
                ),
            )
        }

        onProgress(
            BenchmarkProgressUpdate(
                completedCount = total,
                totalCount = total,
                currentScenario = "Done",
                currentModel = "",
            ),
        )

        return BenchmarkRunResult(
            results = results,
            skippedCategories = preflightResult.skippedCategories,
        )
    }
}
